lib/chat_models/chat_anthropic.ex

```diff
@@ -455,107 +437,72 @@ defmodule LangChain.ChatModels.ChatAnthropic do
-    #
-    #     event: ping
-    #     data: {"type": "ping"}
-    #
-    #     event: content_block_delta
-    #     data: {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "Hello"}}
-    #
-    #     event: content_block_delta
-    #     data: {"type": "content_block_delta", "index": 0, "delta": {"type": "text_delta", "text": "!"}}
-    #
-    #     event: content_block_stop
-    #     data: {"type": "content_block_stop", "index": 0}
-    #
-    #     event: message_delta
-    #     data: {"type": "message_delta", "delta": {"stop_reason": "end_turn", "stop_sequence":null, "usage":{"output_tokens": 15}}}
-    #
-    #     event: message_stop
-    #     data: {"type": "message_stop"}
-    #
-    #
-    # See Anthropic documentation: https://docs.anthropic.com/claude/reference/messages-streaming
-
-    def decode(chunk, buffer) do
-      ((buffer || "") <> chunk)
-      |> String.split("\n\n", trim: true)
-      |> Enum.reduce({[], ""}, fn raw_event, {parsed_data_list, incomplete} ->
-        case decode_event(raw_event) do
-          :ignore ->
-            {parsed_data_list, incomplete}
-
-          {:complete, parsed_data} ->
-            {parsed_data_list ++ [parsed_data], incomplete}
-
-          # If we assume Anthropic is not sending corrupt events, then this can only
-          # ever happen with the last item in the list resulting from the "\n\n" split.
-          {:incomplete, remaining} ->
-            {parsed_data_list, remaining}
-        end
-      end)
-    end
-
-    defp decode_event("event: content_block_delta\ndata: " <> data = raw_event) do
-      parse_event_data(data, raw_event)
-    end
-
-    defp decode_event("event: content_block_start\ndata: " <> data = raw_event) do
-      parse_event_data(data, raw_event)
-    end
-
-    defp decode_event("event: message_delta\ndata: " <> data = raw_event) do
-      parse_event_data(data, raw_event)
-    end
-
-    defp decode_event("event: message_start\n" <> _rest) do
-      :ignore
-    end
-
-    defp decode_event("event: ping\n" <> _rest) do
-      :ignore
-    end
+  @doc false
+  def decode_stream({chunk, buffer}) do
+    # Combine the incoming data with the buffered incomplete data
+    combined_data = buffer <> chunk
+    # Split data by double newline to find complete messages
+    entries = String.split(combined_data, "\n\n", trim: true)
+
+    # The last part may be incomplete if it doesn't end with "\n\n"
+    {to_process, incomplete} =
+      if String.ends_with?(combined_data, "\n\n") do
+        {entries, ""}
+      else
+        # process all but the last, keep the last as incomplete
+        {Enum.slice(entries, 0..-2//1), List.last(entries)}
+      end
 
-    defp decode_event("event: content_block_stop\n" <> _rest) do
-      :ignore
-    end
+    processed =
+      to_process
+      # Trim whitespace from each line
+      |> Stream.map(&String.trim/1)
```

Comment:
Was the thinking behind this implementation over the previous one using pattern matching to be more robust to whitespace?

Reply:
I wasn't sure how large the chunks might be, but it felt like Streams were a good fit. I've profiled Streams vs Enum approaches before and find them very comparable in performance on the small end. Streams have the greatest benefit when the data is very large.

Really, my goal was to may it easier to read and reason about and to make pattern matching more heavily featured.